// @stylable/module-utils/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateStylableJSModuleSource = exports.generateDTSSourceMap = exports.generateDTSContent = exports.stylableModuleFactory = exports.generateModuleSource = exports.createModuleSource = void 0;
/**@deprecated */
var module_source_1 = require("./module-source");
Object.defineProperty(exports, "createModuleSource", { enumerable: true, get: function () { return module_source_1.createModuleSource; } });
Object.defineProperty(exports, "generateModuleSource", { enumerable: true, get: function () { return module_source_1.generateModuleSource; } });
var module_factory_1 = require("./module-factory");
Object.defineProperty(exports, "stylableModuleFactory", { enumerable: true, get: function () { return module_factory_1.stylableModuleFactory; } });
var generate_dts_1 = require("./generate-dts");
Object.defineProperty(exports, "generateDTSContent", { enumerable: true, get: function () { return generate_dts_1.generateDTSContent; } });
var generate_dts_sourcemaps_1 = require("./generate-dts-sourcemaps");
Object.defineProperty(exports, "generateDTSSourceMap", { enumerable: true, get: function () { return generate_dts_sourcemaps_1.generateDTSSourceMap; } });
// ToDo(major): moved to core
var core_1 = require("@stylable/core");
Object.defineProperty(exports, "generateStylableJSModuleSource", { enumerable: true, get: function () { return core_1.generateStylableJSModuleSource; } });
