// @stylable/module-utils/dist/module-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stylableModuleFactory = void 0;
const core_1 = require("@stylable/core");
function stylableModuleFactory(stylableOptions, { runtimePath, runtimeStylesheetId = 'module', injectCSS = true, staticImports = [], moduleType = 'cjs', runtimeId = '0', } = {}) {
    const stylable = new core_1.Stylable(stylableOptions);
    return function stylableToModule(source, path) {
        const { meta, exports } = stylable.transform(stylable.analyze(path, source));
        return (0, core_1.generateStylableJSModuleSource)({
            moduleType: moduleType,
            imports: staticImports.map((from) => ({ from })),
            jsExports: exports,
            namespace: meta.namespace,
            runtimeRequest: runtimePath,
            varType: 'var',
        }, {
            id: runtimeStylesheetId === 'module' ? undefined : meta.namespace,
            css: injectCSS ? meta.targetAst.toString() : '',
            depth: -1,
            runtimeId,
        });
    };
}
exports.stylableModuleFactory = stylableModuleFactory;
