// @stylable/cli/dist/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDefaultLogger = exports.createLogger = exports.levels = void 0;
exports.levels = {
    debug: Symbol('debug'),
    info: Symbol('info'),
    clear: Symbol('clear'),
};
function createLogger(onLog, onClear) {
    return function log(...messages) {
        const clear = messages[messages.length - 1] === exports.levels.clear;
        if (clear) {
            onClear();
            return;
        }
        const info = messages[messages.length - 1] === exports.levels.info;
        const debug = messages[messages.length - 1] === exports.levels.debug;
        if (info || debug) {
            messages.pop();
        }
        onLog(info ? 'info' : 'debug', ...messages);
    };
}
exports.createLogger = createLogger;
function createDefaultLogger() {
    return createLogger((level, ...messages) => level === 'info' && console.log(...messages), console.clear);
}
exports.createDefaultLogger = createDefaultLogger;
