// @stylable/cli/dist/code-mods/st-global-custom-property-to-at-property.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stGlobalCustomPropertyToAtProperty = void 0;
const core_1 = require("@stylable/core");
const index_internal_1 = require("@stylable/core/dist/index-internal");
const stGlobalCustomPropertyToAtProperty = ({ ast, diagnostics, postcss }) => {
    let changed = false;
    ast.walkAtRules('st-global-custom-property', (atRule) => {
        const properties = parseStGlobalCustomProperty(atRule, diagnostics);
        if (!diagnostics.reports.length) {
            for (const property of properties) {
                atRule.before(postcss.atRule({
                    name: 'property',
                    params: `st-global(${property.name})`,
                }));
            }
            atRule.remove();
            changed = true;
        }
    });
    return {
        changed,
    };
};
exports.stGlobalCustomPropertyToAtProperty = stGlobalCustomPropertyToAtProperty;
function parseStGlobalCustomProperty(atRule, diagnostics) {
    const cssVars = [];
    const cssVarsByComma = atRule.params.split(',');
    const cssVarsBySpacing = atRule.params
        .trim()
        .split(/\s+/g)
        .filter((s) => s !== ',');
    if (cssVarsBySpacing.length > cssVarsByComma.length) {
        diagnostics.report(index_internal_1.CSSCustomProperty.diagnostics.GLOBAL_CSS_VAR_MISSING_COMMA(atRule.params), {
            node: atRule,
            word: atRule.params,
        });
        return cssVars;
    }
    for (const entry of cssVarsByComma) {
        const cssVar = entry.trim();
        if ((0, core_1.validateCustomPropertyName)(cssVar)) {
            cssVars.push({
                _kind: 'cssVar',
                name: cssVar,
                global: true,
                alias: undefined,
            });
        }
        else {
            diagnostics.report(index_internal_1.CSSCustomProperty.diagnostics.ILLEGAL_GLOBAL_CSS_VAR(cssVar), {
                node: atRule,
                word: cssVar,
            });
        }
    }
    return cssVars;
}
