// @stylable/cli/dist/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorMessages = exports.buildMessages = void 0;
exports.buildMessages = {
    EMIT_BUNDLE(filePath, fileCount) {
        return `Emitting bundle to "${filePath}" contains ${fileCount} files.`;
    },
    START_WATCHING() {
        return 'Start watching...';
    },
    STOP_WATCHING() {
        return 'Stop watching.';
    },
    CONTINUE_WATCH() {
        return `Watching files...`;
    },
    FINISHED_PROCESSING(count, location) {
        return `Finished processing ${count} ${count === 1 ? 'file' : 'files'}${location ? ` in "${location}"` : ''}.`;
    },
    BUILD_PROCESS_INFO(location) {
        return `Processing files of "${location}"`;
    },
    BUILD_SKIPPED(identifier) {
        return `No stylable files found. build skipped${identifier ? ` for "${identifier}"` : ''}.`;
    },
    SKIP_GENERATED_FILE(location) {
        return `Skipping generated file build of "${location}".`;
    },
    CHANGE_EVENT_TRIGGERED(location) {
        return `Change event triggered for "${location}".`;
    },
    CHANGE_DETECTED(location) {
        return `Change detected at "${location}". Start Processing...`;
    },
    WATCH_SUMMARY(changes, deleted) {
        return `Processed ${changes} changes and ${deleted} deletions.`;
    },
    NO_DIAGNOSTICS() {
        return `Found 0 diagnostics.`;
    },
};
exports.errorMessages = {
    STYLABLE_PROCESS(filePath) {
        return `Stylable failed to process "${filePath}"`;
    },
};
