// @tokey/css-value-parser/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseValueSyntax = exports.margin = exports.background = exports.defineProperty = exports.stringifyCSSValue = exports.parseCSSValue = void 0;
__exportStar(require("./ast-types"), exports);
var value_parser_1 = require("./value-parser");
Object.defineProperty(exports, "parseCSSValue", { enumerable: true, get: function () { return value_parser_1.parseCSSValue; } });
var value_stringify_1 = require("./value-stringify");
Object.defineProperty(exports, "stringifyCSSValue", { enumerable: true, get: function () { return value_stringify_1.stringifyCSSValue; } });
var define_property_1 = require("./define-property");
Object.defineProperty(exports, "defineProperty", { enumerable: true, get: function () { return define_property_1.defineProperty; } });
// properties
var background_1 = require("./properties/background");
Object.defineProperty(exports, "background", { enumerable: true, get: function () { return background_1.background; } });
var margin_1 = require("./properties/margin");
Object.defineProperty(exports, "margin", { enumerable: true, get: function () { return margin_1.margin; } });
// syntax
var value_syntax_parser_1 = require("./value-syntax-parser");
Object.defineProperty(exports, "parseValueSyntax", { enumerable: true, get: function () { return value_syntax_parser_1.parseValueSyntax; } });
