// @stylable/core/dist/features/feature.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFeature = exports.defaultFeatureFlags = void 0;
exports.defaultFeatureFlags = {
    strictCustomProperty: false,
};
const defaultHooks = {
    metaInit() {
        /**/
    },
    analyzeInit() {
        /**/
    },
    analyzeAtRule() {
        /**/
    },
    analyzeSelectorNode() {
        /**/
    },
    analyzeSelectorDone() {
        /**/
    },
    analyzeDeclaration() {
        /**/
    },
    analyzeDone() {
        /**/
    },
    prepareAST() {
        /**/
    },
    transformInit() {
        /**/
    },
    transformResolve() {
        return {};
    },
    transformAtRuleNode() {
        /**/
    },
    transformSelectorNode() {
        /**/
    },
    transformDeclaration() {
        /**/
    },
    transformValue() {
        /**/
    },
    transformJSExports() {
        /**/
    },
    transformLastPass() {
        /**/
    },
};
function createFeature(hooks) {
    return {
        ...defaultHooks,
        ...hooks,
    };
}
exports.createFeature = createFeature;
