// @stylable/core/dist/features/st-custom-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTemplateState = exports.resolveStateParam = exports.sysValidationErrors = exports.systemValidators = exports.createStateWithParamClassName = exports.createBooleanStateClassName = exports.delimiters = exports.transformPseudoClassToCustomState = exports.parseStateValue = exports.parsePseudoStates = exports.hooks = exports.diagnostics = void 0;
const feature_1 = require("./feature");
const custom_state_1 = require("../helpers/custom-state");
Object.defineProperty(exports, "parsePseudoStates", { enumerable: true, get: function () { return custom_state_1.parsePseudoStates; } });
Object.defineProperty(exports, "parseStateValue", { enumerable: true, get: function () { return custom_state_1.parseStateValue; } });
Object.defineProperty(exports, "transformPseudoClassToCustomState", { enumerable: true, get: function () { return custom_state_1.transformPseudoClassToCustomState; } });
Object.defineProperty(exports, "createBooleanStateClassName", { enumerable: true, get: function () { return custom_state_1.createBooleanStateClassName; } });
Object.defineProperty(exports, "createStateWithParamClassName", { enumerable: true, get: function () { return custom_state_1.createStateWithParamClassName; } });
Object.defineProperty(exports, "systemValidators", { enumerable: true, get: function () { return custom_state_1.systemValidators; } });
Object.defineProperty(exports, "sysValidationErrors", { enumerable: true, get: function () { return custom_state_1.validationErrors; } });
Object.defineProperty(exports, "resolveStateParam", { enumerable: true, get: function () { return custom_state_1.resolveStateParam; } });
Object.defineProperty(exports, "isTemplateState", { enumerable: true, get: function () { return custom_state_1.isTemplateState; } });
exports.diagnostics = {
    ...custom_state_1.stateDiagnostics,
};
// HOOKS
exports.hooks = (0, feature_1.createFeature)({});
const delimiters = {
    booleanStateDelimiter: custom_state_1.booleanStateDelimiter,
    stateMiddleDelimiter: custom_state_1.stateMiddleDelimiter,
    stateWithParamDelimiter: custom_state_1.stateWithParamDelimiter,
};
exports.delimiters = delimiters;
